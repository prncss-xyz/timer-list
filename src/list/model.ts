import { z } from "zod";

export const itemSchema = z.object({ seconds: z.number() });
export type Item = z.infer<typeof itemSchema>;
export const nullItem: Item = { seconds: 0 };

export const itemsSchema = z.array(itemSchema);
