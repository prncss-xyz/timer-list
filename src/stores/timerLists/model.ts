import { z } from "zod";

import { getUUID } from "@/utils/uuid";

const itemSchema = z.object({
  id: z.string(),
  seconds: z.number(),
});
export type Item = z.infer<typeof itemSchema>;
export function getNullItem() {
  return { seconds: 0, id: getUUID() };
}

const timerListSchema = z.object({
  index: z.number(),
  items: z.array(itemSchema),
});
export type TimerList = z.infer<typeof timerListSchema>;
export function getNullTimerList() {
  return { index: 0, items: [getNullItem()] };
}

export const validateTimerListSchema = (v: unknown) => {
  try {
    return timerListSchema.parse(v);
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
  } catch (_err) {}
};
