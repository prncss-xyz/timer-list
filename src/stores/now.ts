import { atom } from "jotai";

const resolution = 200;

export const nowAtom = atom(0);
nowAtom.onMount = (setAtom) => {
  const timer = setInterval(() => {
    return setAtom(Date.now());
  }, resolution);
  return () => clearInterval(timer);
};
