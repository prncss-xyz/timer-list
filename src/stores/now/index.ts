import { atom } from "jotai";

export const nowAtom = atom(0);
