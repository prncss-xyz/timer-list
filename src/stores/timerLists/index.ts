// we use the plural form in filename as it will eventually manage many lisits
import { atom } from "jotai";
import { atomEffect } from "jotai-effect";
import { focusAtom } from "jotai-optics";

import { Item, normalize } from "./model";

import { definedAtom } from "@/utils/definedAtom";
import { fromSeconds, toSeconds } from "@/utils/seconds";
import { getUUID } from "@/utils/uuid";

/* const rawTimerListAtom = atom(normalize({ index: 0, items: [] })); */
const rawTimerListAtom = atom(
  normalize({
    index: 0,
    items: [
      { seconds: 1, id: getUUID() },
      { seconds: 2, id: getUUID() },
      { seconds: 3, id: getUUID() },
    ],
  }),
);
export const timerListAtom = focusAtom(rawTimerListAtom, (o) =>
  o.rewrite(normalize),
);

export const currentIndexAtom = focusAtom(timerListAtom, (o) =>
  o.prop("index"),
);

export const itemsAtom = focusAtom(timerListAtom, (o) => o.prop("items"));

const currentItemAtom = atom(
  (get) => {
    const { index, items } = get(timerListAtom);
    return items[index];
  },
  (get, set, item: Item) => {
    const lists = get(timerListAtom);
    let { index, items } = lists;
    items = items.toSpliced(index, 1, item);
    set(timerListAtom, { ...lists, items });
  },
);

export const currentIdAtom = atom(
  (get) => {
    return get(currentItemAtom).id;
  },
  (get, set, id: string) => {
    const lists = get(timerListAtom);
    const { items } = lists;
    const index = items.findIndex((item) => item.id === id);
    set(timerListAtom, { ...lists, index, items });
  },
);

export const currentSecondsAtom = focusAtom(currentItemAtom, (o) =>
  o.prop("seconds"),
);

export const getIdItemAtom = (id: string) =>
  definedAtom(
    focusAtom(timerListAtom, (o) =>
      o.prop("items").find((item) => item.id === id),
    ),
    { id: "", seconds: 0 },
  );

export const getIdItemSecondsAtom = (id: string) =>
  focusAtom(getIdItemAtom(id), (o) => o.prop("seconds"));

export const getIdItemSecondsTextAtom = (id: string) =>
  focusAtom(getIdItemSecondsAtom(id), (o) => o.iso(fromSeconds, toSeconds));

export const duplicateIdAtom = atom(null, (get, set, id: string) => {
  const lists = get(timerListAtom);
  let { items, index: currentIndex } = lists;
  const index = items.findIndex((item) => item.id === id);
  if (index < 0) return;
  const item = { ...items[index], id: getUUID() };
  items = items.toSpliced(index, 0, item);
  if (index <= currentIndex) currentIndex++;
  set(timerListAtom, { ...lists, index: currentIndex, items });
});

export const removeIdAtom = atom(null, (get, set, id: string) => {
  const lists = get(timerListAtom);
  let { items, index: currentIndex } = lists;
  const index = items.findIndex((item) => item.id === id);
  if (index < 0) return;
  items = items.toSpliced(index, 1);
  if (index < currentIndex) currentIndex--;
  set(timerListAtom, { ...lists, index: currentIndex, items });
});

export const clearItemsAtom = atom(null, (_get, set) => {
  set(itemsAtom, []);
});

export const getTimerUpdateEffect = (cb: () => void) =>
  atomEffect((get) => {
    get(currentSecondsAtom);
    get(currentIdAtom);
    cb();
  });
