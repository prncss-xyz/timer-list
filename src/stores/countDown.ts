import { atom, useAtomValue } from "jotai";
import { atomEffect } from "jotai-effect";
import { useMemo } from "react";

import { currentIndexAtom, currentSecondsAtom } from "./timerLists";
import {
  resetTimerAtom,
  stopTimerAtom,
  timerActiveAtom,
  timerElapsedAtom,
} from "./timers";

import { fromSeconds } from "@/utils/seconds";

const countDownMsAtom = atom((get) => {
  const elapsed = get(timerElapsedAtom);
  const total = get(currentSecondsAtom) * 1000;
  return Math.max(0, total - elapsed);
});

export const countDownSecondsAtom = atom((get) =>
  Math.ceil(get(countDownMsAtom) / 1000),
);

export const getAlarmEffect = (alarm: () => void) =>
  atomEffect((get, set) => {
    if (get(timerActiveAtom) === false || get(countDownMsAtom) > 0) return;
    set(stopTimerAtom);
    set(currentIndexAtom, (index) => index + 1);
    set(resetTimerAtom);
    alarm();
  });

export function useInitCountDown(alarm: () => void) {
  const effect = useMemo(() => getAlarmEffect(alarm), [alarm]);
  useAtomValue(effect);
}

export const countDownTextAtom = atom((get) =>
  fromSeconds(get(countDownSecondsAtom)),
);
