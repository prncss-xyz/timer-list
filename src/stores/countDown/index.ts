import { atom } from "jotai";
import { atomEffect } from "jotai-effect";

import { activeSecondsAtom, nextActiveItemAtom } from "../timerLists";
import { timerElapsedAtom, stopTimerAtom, resetTimerAtom } from "../timers";

import { fromSeconds } from "@/utils/seconds";

function getCountDown(elapsedMs: number, total: number | undefined) {
  if (total === undefined) return 0;
  return Math.max(0, Math.ceil(total - elapsedMs / 1000));
}

const countDownAtom = atom((get) => {
  return getCountDown(get(timerElapsedAtom), get(activeSecondsAtom));
});

export const countDownTextAtom = atom((get) => fromSeconds(get(countDownAtom)));

export const getAlarmEffect = (alarm: () => void) =>
  atomEffect((get, set) => {
    if (get(countDownAtom) > 0) return;
    set(stopTimerAtom);
    set(nextActiveItemAtom);
    set(resetTimerAtom);
    alarm();
  });
