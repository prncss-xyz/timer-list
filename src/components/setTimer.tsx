import { rewrite } from "@constellar/core";
import { focusAtom } from "@constellar/jotai";
import Ionicons from "@expo/vector-icons/Ionicons";
import { router } from "expo-router";
import { Provider, atom, useAtom, useAtomValue, useSetAtom } from "jotai";
import { useHydrateAtoms } from "jotai/utils";
import React, { ReactNode, useCallback, useMemo } from "react";
import { Pressable, Text, View } from "react-native";

import { WideButton } from "./wideButton";

import { HeadSeparator } from "@/components/headSeparator";
import { TimerView } from "@/components/timerView";
import { useColor } from "@/hooks/color";
import { getIdItemSecondsTextAtom } from "@/stores/timerLists";
import { borderWidths, fontSizes, sizes, spaces, styles } from "@/styles";
import { normalizeSeconds } from "@/utils/seconds";

const rawTextAtom = atom("");
const textAtom = focusAtom(rawTextAtom, rewrite(normalizeSeconds));
const updateAtom = atom(null, (get, _set, cb: (text: string) => void) => {
  cb(get(textAtom));
});

function Hydrator({ text, children }: { text: string; children: ReactNode }) {
  useHydrateAtoms([[rawTextAtom, text]]);
  return children;
}

function appendText(text: string) {
  return (str: string) => str + text;
}

function backspaceText(str: string) {
  return str.slice(0, -1);
}

function Timer() {
  const text = useAtomValue(textAtom);
  return (
    <View aria-label="timer">
      <TimerView color={useColor("brand")} text={text} />
    </View>
  );
}

function Cancel() {
  return (
    <Pressable
      aria-label="cancel"
      onPress={router.back}
      style={styles.iconPlace}
    >
      <Ionicons color={useColor("brand")} name="close" size={sizes.icon} />
    </Pressable>
  );
}

function ListBar() {
  return (
    <View
      style={{
        flexDirection: "row",
        justifyContent: "space-between",
        width: "100%",
        alignItems: "flex-start",
        padding: spaces[5],
        paddingBottom: 0,
      }}
    >
      <View style={{ width: sizes.icon }} />
      <Timer />
      <Cancel />
    </View>
  );
}

function Digit({ text }: { text: string }) {
  const setText = useSetAtom(textAtom);
  const onPress = useCallback(() => setText(appendText(text)), [setText, text]);
  const brand = useColor("brand");
  return (
    <Pressable
      aria-label={`digit ${text}`}
      onPress={onPress}
      style={{
        width: sizes.digit,
        height: sizes.digit,
        flexDirection: "row",
        justifyContent: "center",
        alignItems: "center",
        borderColor: brand,
        borderStyle: "solid",
        borderWidth: borderWidths.light,
      }}
    >
      <Text style={[styles.mono500, { fontSize: fontSizes[20], color: brand }]}>
        {text}
      </Text>
    </Pressable>
  );
}

function Backspace() {
  const setText = useSetAtom(textAtom);
  const onPress = useCallback(() => setText(backspaceText), [setText]);
  return (
    <Pressable
      onPress={onPress}
      aria-label="backspace"
      style={{
        width: sizes.digit,
        height: sizes.digit,
        flexDirection: "row",
        justifyContent: "center",
        alignItems: "center",
      }}
    >
      <Ionicons color={useColor("brand")} name="backspace" size={sizes.digit} />
    </Pressable>
  );
}

function Done({ setText }: { setText: (text: string) => void }) {
  const update = useSetAtom(updateAtom);
  const onPress = useCallback(() => {
    update(setText);
    router.back();
  }, [setText, update]);
  const active = useColor("active");
  return <WideButton onPress={onPress} color={active} text="done" />;
}

function Grid() {
  const gap = spaces[15];
  return (
    <View
      style={{
        flexDirection: "row",
        flexWrap: "wrap",
        gap,
        maxWidth: 2 * gap + 3 * sizes.digit,
      }}
    >
      <Digit text="1" />
      <Digit text="2" />
      <Digit text="3" />
      <Digit text="4" />
      <Digit text="5" />
      <Digit text="6" />
      <Digit text="7" />
      <Digit text="8" />
      <Digit text="9" />
      <Digit text="00" />
      <Digit text="0" />
      <Backspace />
    </View>
  );
}

export function SetTimer({ timerId }: { timerId: string }) {
  const [text, setText] = useAtom(
    useMemo(() => getIdItemSecondsTextAtom(timerId), [timerId]),
  );
  if (text === undefined) return null;
  return (
    <Provider>
      {/* we can only pass serializable values, hence setText is not passed as an atom here, but as a prop later on */}
      <Hydrator text={text}>
        <ListBar />
        <HeadSeparator />
        <View
          style={{
            flex: 1,
            alignItems: "center",
            justifyContent: "center",
          }}
        >
          <Grid />
        </View>
        <Done setText={setText} />
      </Hydrator>
    </Provider>
  );
}
