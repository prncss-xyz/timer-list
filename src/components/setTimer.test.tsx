import {
  fireEvent,
  render,
  screen,
  within,
} from "@testing-library/react-native";
import { router } from "expo-router";
import { createStore, Provider } from "jotai";

import { SetTimer } from "./setTimer";

import { timerListAtom } from "@/stores/timerLists";
import { TimerList } from "@/stores/timerLists/model";

jest.mock("expo-router", () => ({
  router: { back: jest.fn() },
}));

describe("set-timer", () => {
  it("should rendrer properly", () => {
    const store = createStore();
    store.set(timerListAtom, {
      index: 0,
      items: [{ seconds: 1, id: "a" }],
    });
    const tree = render(
      <Provider store={store}>
        <SetTimer timerId="a" />
      </Provider>,
    );
    expect(tree).toMatchSnapshot();
  });
  it("should append digits to timer", () => {
    const timerList: TimerList = {
      index: 0,
      items: [{ seconds: 1, id: "a" }],
    };
    for (let i = 0; i < 10; ++i) {
      const digit = String(i);
      const store = createStore();
      store.set(timerListAtom, timerList);
      render(
        <Provider store={store}>
          <SetTimer timerId="a" />
        </Provider>,
      );
      // start playing
      fireEvent.press(screen.getByText(digit));
      within(screen.getByLabelText("timer")).getByText(`00:00:1${digit}`);
    }
  });
  it("should append 00 to timer", () => {
    const store = createStore();
    store.set(timerListAtom, {
      index: 0,
      // 23 * 3600 + 1 => "23:00:01"
      items: [{ seconds: 23 * 3600 + 1, id: "a" }],
    });
    render(
      <Provider store={store}>
        <SetTimer timerId="a" />
      </Provider>,
    );
    fireEvent.press(screen.getByText("00"));
    within(screen.getByLabelText("timer")).getByText("00:01:00");
  });
  it("should close and not update value", () => {
    jest.resetAllMocks();
    const store = createStore();
    store.set(timerListAtom, {
      index: 0,
      items: [{ seconds: 1, id: "a" }],
    });
    render(
      <Provider store={store}>
        <SetTimer timerId="a" />
      </Provider>,
    );
    fireEvent.press(screen.getByText("1"));
    fireEvent.press(screen.getByLabelText("cancel"));
    expect(store.get(timerListAtom).items).toEqual([{ seconds: 1, id: "a" }]);
    expect((router.back as any).mock.calls).toHaveLength(1);
  });
  it("should close and update value", () => {
    jest.resetAllMocks();
    const store = createStore();
    store.set(timerListAtom, {
      index: 0,
      items: [{ seconds: 1, id: "a" }],
    });
    render(
      <Provider store={store}>
        <SetTimer timerId="a" />
      </Provider>,
    );
    fireEvent.press(screen.getByText("1"));
    fireEvent.press(screen.getByLabelText("done"));
    expect(store.get(timerListAtom).items).toEqual([{ seconds: 11, id: "a" }]);
    expect((router.back as any).mock.calls).toHaveLength(1);
  });
  it("should remove last character to timer", () => {
    const store = createStore();
    store.set(timerListAtom, {
      index: 0,
      // 23 * 3600 + 1 => "12:00:34"
      items: [{ seconds: 12 * 3600 + 34, id: "a" }],
    });
    render(
      <Provider store={store}>
        <SetTimer timerId="a" />
      </Provider>,
    );
    // start playing
    fireEvent.press(screen.getByLabelText("backspace"));
    within(screen.getByLabelText("timer")).getByText("01:20:03");
  });
});
