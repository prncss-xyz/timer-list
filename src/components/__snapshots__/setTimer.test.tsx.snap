// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`set-timer should rendrer properly 1`] = `
[
  <View
    style={
      {
        "alignItems": "flex-start",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 5,
        "paddingBottom": 0,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "width": 32,
        }
      }
    />
    <View
      aria-label="timer"
    >
      <Text
        style={
          [
            {
              "borderColor": "#0090FF",
              "color": "#0090FF",
            },
            {
              "borderStyle": "solid",
              "borderWidth": 1,
              "justifyContent": "center",
              "padding": 10,
              "textAlign": "center",
            },
            {
              "fontFamily": "RobotoMono-Medium",
              "fontWeight": "500",
            },
            {
              "fontSize": 28,
            },
          ]
        }
      >
        00:00:01
      </Text>
    </View>
    <View
      accessibilityLabel="cancel"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "height": 32,
          "width": 32,
        }
      }
    >
      <Text />
    </View>
  </View>,
  <View
    style={
      {
        "borderBottomWidth": 1,
        "borderColor": "#0090FF",
        "borderStyle": "solid",
        "height": 35,
      }
    }
  />,
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "flexWrap": "wrap",
          "gap": 15,
          "maxWidth": 180,
        }
      }
    >
      <View
        accessibilityLabel="digit 1"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "borderColor": "#0090FF",
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "width": 50,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "RobotoMono-Medium",
                "fontWeight": "500",
              },
              {
                "color": "#0090FF",
                "fontSize": 20,
              },
            ]
          }
        >
          1
        </Text>
      </View>
      <View
        accessibilityLabel="digit 2"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "borderColor": "#0090FF",
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "width": 50,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "RobotoMono-Medium",
                "fontWeight": "500",
              },
              {
                "color": "#0090FF",
                "fontSize": 20,
              },
            ]
          }
        >
          2
        </Text>
      </View>
      <View
        accessibilityLabel="digit 3"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "borderColor": "#0090FF",
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "width": 50,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "RobotoMono-Medium",
                "fontWeight": "500",
              },
              {
                "color": "#0090FF",
                "fontSize": 20,
              },
            ]
          }
        >
          3
        </Text>
      </View>
      <View
        accessibilityLabel="digit 4"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "borderColor": "#0090FF",
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "width": 50,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "RobotoMono-Medium",
                "fontWeight": "500",
              },
              {
                "color": "#0090FF",
                "fontSize": 20,
              },
            ]
          }
        >
          4
        </Text>
      </View>
      <View
        accessibilityLabel="digit 5"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "borderColor": "#0090FF",
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "width": 50,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "RobotoMono-Medium",
                "fontWeight": "500",
              },
              {
                "color": "#0090FF",
                "fontSize": 20,
              },
            ]
          }
        >
          5
        </Text>
      </View>
      <View
        accessibilityLabel="digit 6"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "borderColor": "#0090FF",
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "width": 50,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "RobotoMono-Medium",
                "fontWeight": "500",
              },
              {
                "color": "#0090FF",
                "fontSize": 20,
              },
            ]
          }
        >
          6
        </Text>
      </View>
      <View
        accessibilityLabel="digit 7"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "borderColor": "#0090FF",
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "width": 50,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "RobotoMono-Medium",
                "fontWeight": "500",
              },
              {
                "color": "#0090FF",
                "fontSize": 20,
              },
            ]
          }
        >
          7
        </Text>
      </View>
      <View
        accessibilityLabel="digit 8"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "borderColor": "#0090FF",
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "width": 50,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "RobotoMono-Medium",
                "fontWeight": "500",
              },
              {
                "color": "#0090FF",
                "fontSize": 20,
              },
            ]
          }
        >
          8
        </Text>
      </View>
      <View
        accessibilityLabel="digit 9"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "borderColor": "#0090FF",
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "width": 50,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "RobotoMono-Medium",
                "fontWeight": "500",
              },
              {
                "color": "#0090FF",
                "fontSize": 20,
              },
            ]
          }
        >
          9
        </Text>
      </View>
      <View
        accessibilityLabel="digit 00"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "borderColor": "#0090FF",
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "width": 50,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "RobotoMono-Medium",
                "fontWeight": "500",
              },
              {
                "color": "#0090FF",
                "fontSize": 20,
              },
            ]
          }
        >
          00
        </Text>
      </View>
      <View
        accessibilityLabel="digit 0"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "borderColor": "#0090FF",
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "width": 50,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "RobotoMono-Medium",
                "fontWeight": "500",
              },
              {
                "color": "#0090FF",
                "fontSize": 20,
              },
            ]
          }
        >
          0
        </Text>
      </View>
      <View
        accessibilityLabel="backspace"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "width": 50,
          }
        }
      >
        <Text />
      </View>
    </View>
  </View>,
  <View
    accessibilityLabel="done"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "padding": 5,
      }
    }
  >
    <Text
      style={
        [
          {
            "borderWidth": 1,
            "fontFamily": "Inter-SemiBold",
            "fontSize": 20,
            "fontWeight": "600",
            "padding": 5,
            "textAlign": "center",
            "textTransform": "uppercase",
          },
          {
            "borderColor": "#3DD68C",
            "color": "#3DD68C",
          },
        ]
      }
    >
      done
    </Text>
  </View>,
]
`;
